passenger_car_dynamic_model/src/passenger_car_dynamic_model/PassengerCarDynamicModel.cpp

```diff
@@ -0,0 +1,283 @@
+
+  // Check if all the required parameters could be loaded
+  if (!(l_f_param && l_r_param && R_f_param && R_r_param
+         && long_stiff_param && lat_stiff_param && inertia_param && mass_param)) {
+
+    std::ostringstream msg;
+    msg << "One of the required parameters could not be found or read " 
+      << " length_to_f: " << l_f_param 
+      << " length_to_r: " << l_r_param 
+      << " effective_wheel_radius_f: " << R_f_param 
+      << " effective_wheel_radius_r: " << R_r_param 
+      << " tire_longitudinal_stiffness: " << long_stiff_param
+      << " tire_cornering_stiffness: " << lat_stiff_param 
+      << " moment_of_inertia: " << inertia_param
+      << " vehicle_mass: " << mass_param;
+
+    throw std::invalid_argument(msg.str());
+  }
+}
+
+std::vector<VehicleState> PassengerCarDynamicModel::predict(const VehicleState& initial_state,
+  double timestep, double delta_t) {
+
+    // Populate control inputs
+    // This predict function takes in no new control inputs so extract the old ones from the state vector
+    VehicleControlInput control_input;
+    control_input.target_steering_angle = initial_state.prev_steering_cmd;
+    control_input.target_velocity = initial_state.prev_vel_cmd;
+
+    // Ensure we run at least 1 step
+    size_t num_steps;
+    if (delta_t <= timestep) {
+      num_steps = 1;
+    } else {
+      num_steps = delta_t / timestep;
+    }
+
+    std::vector<VehicleControlInput> control_inputs(num_steps, control_input);
+
+    // Call default predict method
+    return predict(initial_state, control_inputs, timestep);
+  }
+
+std::vector<VehicleState> PassengerCarDynamicModel::predict(const VehicleState& initial_state,
+  const std::vector<VehicleControlInput>& control_inputs, double timestep) {
+        
+    // Construct output vector
+    std::vector<VehicleState> resulting_states;
+    resulting_states.reserve(control_inputs.size());
+
+    // Construct ode output vector
+    std::vector<std::tuple<double, ODESolver::State>> ode_outputs;
+    ode_outputs.reserve(control_inputs.size());
+
+    // Populate initial condition
+    ODESolver::State state(ODE_STATE_SIZE, 0);
+    state[0]  = initial_state.X_pos_global;
+    state[1]  = initial_state.Y_pos_global;
+    state[2]  = initial_state.orientation;
+    state[3]  = initial_state.longitudinal_vel;
+    state[4]  = initial_state.lateral_vel;
+    state[5]  = initial_state.yaw_rate;
+    state[6]  = initial_state.front_wheel_rotation_rate;
+    state[7]  = initial_state.rear_wheel_rotation_rate;
+    state[8]  = initial_state.steering_angle;
+
+    // Integrate ODE
+    ODESolver::rk4<VehicleControlInput>(
+      ode_func_,
+      control_inputs.size(),
+      timestep,
+      state,
+      control_inputs,
+      ode_outputs,
+      post_step_func_
+    );
+
+    // Convert result to target output
+    for (size_t j = 0; j < ode_outputs.size(); j++) {
+      const ODESolver::State new_state = std::get<1>(ode_outputs[j]);
```

Comment:
If we throw away the timestamp for each ODESolver::State, why do we want to include it in ode_outputs?

Reply:
It was a generalization for the ODESolver. Having t included means the data could be easily saved and graphed.